### PR TITLE
Keep trailing slash for datasource proxy requests

### LINKED ISSUE #13287: Datasource proxy is stripping the trailing slash
I am building a datasource to read data directly from a simple HTTP server [file-system-datasource](https://github.com/ryantxu/file-system-datasource)

When requesting a folder:
`http://host/path/folder/`
The proxy request in the browser is:
`/api/datasources/proxy/27/folder/`

This shows up in the NGINX access.log as:
`"GET /data/store/folder HTTP/1.1" 301 194 ...`  << note the missing slash after folder
and sends a redirect to the the original (not proxied) url:
`http://host/path/folder/`

I don't see anything obvious in:
https://github.com/grafana/grafana/blob/master/pkg/api/pluginproxy/ds_proxy.go#L71

Any pointers on what to even start looking at would be great

### PR DESCRIPTION
**Pull request by ryantxu**
Fixes #13287

The [datasource proxy route](https://github.com/grafana/grafana/blob/master/pkg/api/api.go#L255) is registered with macron using:
```
apiRoute.Any("/datasources/proxy/:id/*", reqSignedIn, hs.ProxyDataSourceRequest)
```
And the path is passed to a DataSourceProxy from:
```
proxyPath := c.Params("*")
```
It looks like macron does not include the trailing slash in its route parsing.

This patch checks the URL directly and appends the trailing slash.  (there may be a better way!)

Is there an existing test I could copy to that uses the routing?

### CODE REVIEW

**Comment by marefr** [comment]
@ryantxu your problem is what - that your server requires routes to end with a slash?

You can look at using `fakeReqWithParams` with a fake `handlerFunc` to verify the route/params: https://github.com/grafana/grafana/blob/master/pkg/api/dashboard_test.go#L249-L250

File: pkg/api/dataproxy.go [hunk 1]
```diff
@@ -54,6 +54,15 @@ func (hs *HTTPServer) ProxyDataSourceRequest(c *m.ReqContext) {
 	}
 
 	proxyPath := c.Params("*")
+
+	// Check for a trailing slash
+	if len(proxyPath) > 1 {
+		path := c.Req.URL.Path
+		if path[len(path)-1] == '/' && path[len(path)-2] != '/' {
```
Comment (torkelo) [inline pkg/api/dataproxy.go]: why are you checking if eiher last or next to last is forward slash, should it not check if last is forward slash in original path, and if it's not forward slash in proxyPath  (that way if it's fixed in macaron it wont introduce a bug here)

Comment (ryantxu) [reply pkg/api/dataproxy.go]: good catch -- i changed it

**Comment by ryantxu** [comment]
> @ryantxu your problem is what - that your server requires routes to end with a slash?

See #13287, I am trying to list files from apache/nginx servers -- when you make a request to a folder without the trailing slash, they redirect to the path with a slash.  Everything works fine with the direct (not proxy) datasource.

**Review by marefr — APPROVED** [review]
LGTM

**Comment by ryantxu** [comment]
Thanks @marefr

**Comment by marefr** [comment]
Thank you for contributing to Grafana!

**Pull request merged**